cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactional.java

```diff
@@ -0,0 +1,370 @@
+    Transaction transaction = txClient.startLong();
+    try {
+      // Create a DatasetContext that uses the newly created transaction
+      TransactionalDatasetContext datasetContext = new TransactionalDatasetContext(transaction, datasetCache,
+                                                                                   asyncCommit);
+      activeDatasetContext.set(datasetContext);
+
+      // Call the runnable
+      runnable.run(datasetContext);
+
+      // Persist the changes
+      datasetContext.flush();
+
+      if (!asyncCommit) {
+        // Commit transaction
+        txClient.commit(datasetContext.getTransaction());
+        activeDatasetContext.remove();
+        datasetContext.postCommit();
+        datasetContext.discardDatasets();
+      }
+    } catch (Throwable t) {
+      // Any exception will cause invalidation of the transaction
+      activeDatasetContext.remove();
+      Transactions.invalidateQuietly(txClient, transaction);
+      Throwables.propagateIfInstanceOf(t, TransactionFailureException.class);
+      throw new TransactionFailureException("Failed to execute TxRunnable in transaction", t);
+    }
+  }
+
+  /**
+   * A {@link DatasetContext} to be used for the transactional execution. All {@link Dataset} instance
+   * created through instance of this class will be using the same transaction.
+   *
+   * Instance of this class is safe to use from multiple threads concurrently. This is for supporting Spark program
+   * with multiple threads that drive computation concurrently within the same transaction.
+   */
+  @ThreadSafe
+  private final class TransactionalDatasetContext implements DatasetContext, TransactionInfo {
+
+    private final Transaction transaction;
+    private final DynamicDatasetCache datasetCache;
+    private final Set<Dataset> datasets;
+    private final Set<Dataset> discardDatasets;
+    private final CountDownLatch completion;
+    private volatile boolean jobStarted;
+
+    private TransactionalDatasetContext(Transaction transaction,
+                                        DynamicDatasetCache datasetCache, boolean asyncCommit) {
+      this.transaction = transaction;
+      this.datasetCache = datasetCache;
+      this.datasets = Collections.synchronizedSet(new HashSet<Dataset>());
+      this.discardDatasets = Collections.synchronizedSet(new HashSet<Dataset>());
+      this.completion = asyncCommit ? new CountDownLatch(1) : null;
+    }
+
+    @Override
+    @Nonnull
+    public Transaction getTransaction() {
+      return transaction;
+    }
+
+    @Override
+    public boolean commitOnJobEnded() {
+      return completion != null;
+    }
+
+    @Override
+    public void onJobStarted() {
+      jobStarted = true;
+    }
+
+    @Override
+    public void onTransactionCompleted(boolean jobSucceeded, @Nullable TransactionFailureException failureCause) {
+      // Shouldn't happen
+      Preconditions.checkState(commitOnJobEnded(), "Not expecting transaction to be completed");
```

Comment:
The contract of the `TransactionInfo.onTransactionCompleted` will be called from the Spark listener (asynchronously) when job ended if and only if `commitOnJobEnded` return `true`.